### PR TITLE
DistributedPCA

### PR DESCRIPTION
**Pull request by ordabayevy**

### CODE REVIEW

**Comment by ordabayevy** [comment]
Blocked by #61

File: scvid/callbacks/distributed_pca.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import lightning.pytorch as pl
+import torch
+import torch.distributed as dist
+
+from scvid.module import IncrementalPCA
+
+
+class DistributedPCA(pl.Callback):
+    """
+    Distributed PCA.
```
Comment (mbabadi) [inline scvid/callbacks/distributed_pca.py:15]: Please write a good amount of documentation about the logic, because it is beautiful! :)

File: scvid/callbacks/distributed_pca.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        i = 0
+        while world_size > 1:
+            # level i
+            # at most two local ranks (leading and trailing)
+            if rank % 2 == 0:  # leading rank
+                if rank < world_size:
+                    # if there is a trailing rank
+                    # then concatenate and merge SVDs
+                    src = (rank + 1) * 2**i
+                    A = torch.diag(S_k) @ V_kg
```
Comment (mbabadi) [inline scvid/callbacks/distributed_pca.py:52]: use `einsum`?

Comment (mbabadi) [inline scvid/callbacks/distributed_pca.py:52]: maybe use more descriptive variables here, like:
- `A` -> `self_X`
- `B` -> `other_X`
- `C` -> `joined_X`
- `m` -> `self_X_size`
- `n`-> `other_X_size`

File: scvid/callbacks/distributed_pca.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                        B_mean = torch.zeros_like(A_mean)
+                        dist.recv(B_mean, src=src)
+                        mean_correction = (
+                            math.sqrt(m * n / (m + n)) * (A_mean - B_mean)[None, :]
+                        )
+                        C = torch.cat([C, mean_correction], dim=0)
+                        x_mean_g = A_mean * m / (m + n) + B_mean * n / (m + n)
+                    _, S_q, V_gq = torch.svd_lowrank(C, q=k + p)
+                    # update parameters
+                    S_k = S_q[:k]
```
Comment (mbabadi) [inline scvid/callbacks/distributed_pca.py:69]: The same comment about working in q-dim space applies here ... let's discuss.

File: scvid/callbacks/distributed_pca.py (hunk 1 above)
Comment (mbabadi) [reply scvid/callbacks/distributed_pca.py:15]: ... and the code is very compact and efficient!  you're crazy.

**Review by mbabadi — CHANGES_REQUESTED** [review]

**Review by mbabadi — CHANGES_REQUESTED** [review]
You use PL like a magician! :)

**Review by mbabadi — APPROVED** [review]
